File: src/main/java/com/zeros/scriptassistant/U2.java
package com.zeros.scriptassistant;

import cn.hutool.core.util.NumberUtil;

import java.awt.*;
import java.io.File;
import java.io.IOException;

public class U2 {
    private PythonExecutor executor;
    private String failMsg;

    public boolean init(String path, String serial, String objectName) {
        try {
            failMsg = "Init failed, python interpreter path: " + path;
            executor = new PythonExecutor(path);
            String errorMsg = executor.getErrorMessage();
            if (errorMsg != null) {
                failMsg = errorMsg;
                return false;
            }
            String result = executor.executeCode("import uiautomator2 as u2");
            if (result.contains("ModuleNotFoundError:")) {
                System.out.println(result);
                failMsg = "please install uiautomator2 first";
                System.out.println(failMsg);
                return false;
            }
            if (serial == null) {
                result = executor.executeCode(objectName + " = u2.connect()");
            } else {
                result = executor.executeCode(objectName + " = u2.connect('" + serial + "')");
            }
            if (result.contains("NameError:") || result.contains("状态异常")) {
                failMsg = result;
                System.out.println(failMsg);
                return false;
            }
        } catch (IOException e) {
            e.printStackTrace();
            return false;
        }
        return true;
    }

    public String getInitFailMsg() {
        return failMsg;
    }

    public String executeCode(String code) {
        return executeCode(code, 0);
    }

    public String executeCode(String code, int indentLevel) {
        if (executor == null) {
            return null;
        }
        try {
            return executor.executeCode(code, indentLevel);
        } catch (IOException e) {
            e.printStackTrace();
        }
        return null;
    }

    public void dumpUI(File screenShotSaveFile, File hierarchySaveFile) {
        // device.dump_ui(dir)
        executeCode("d.screenshot(r'" + screenShotSaveFile.getAbsolutePath() + "')");
        executeCode("with open(r'" + hierarchySaveFile.getAbsolutePath() + "', 'w', encoding=\"utf-8\") as f: f.write(d.dump_hierarchy())");
        executeCode("");
        executeCode("");
        executeCode("");
    }

    public Rectangle getNodeBounds(String code) {
        String result = executeCode("print(" + code + ".info['bounds'])");
        executeCode("");
        if (result == null) {
            return null;
        }
        System.out.println(result);
        // {'bottom': 1555, 'left': 296, 'right': 540, 'top': 1278}
        // 判断格式是否正确
        if (!result.contains("'left': ") || !result.contains("'top': ") || !result.contains("'right': ") || !result.contains("'bottom': ")) {
            return null;
        }
        // 解析结果
        int left = NumberUtil.parseInt(result.split("'left': ")[1].split(",")[0]);
        int top = NumberUtil.parseInt(result.split("'top': ")[1].split("}")[0]);
        int right = NumberUtil.parseInt(result.split("'right': ")[1].split(",")[0]);
        int bottom = NumberUtil.parseInt(result.split("'bottom': ")[1].split(",")[0]);
        return new Rectangle(left, top, right-left, bottom-top);
    }

    public void close() {
        try {
            executor.close();
            executor = null;
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void main(String[] args) {
        new U2().init("F:\\PycharmProjects\\U2Test\\venv\\Scripts\\python.exe", null, "d");
    }
}

File: src/main/kotlin/com/zeros/scriptassistant/U2.kt
package com.zeros.scriptassistant

import cn.hutool.core.util.NumberUtil
import java.awt.Rectangle
import java.io.File
import java.io.IOException

class U2 {
    private var executor: PythonExecutor? = null
    private var failMsg: String? = null

    fun init(path: String?, serial: String?, objectName: String): Boolean {
        return try {
            failMsg = "Init failed, python interpreter path: $path"
            executor = PythonExecutor(path)
            val errorMsg = executor!!.errorMessage
            if (errorMsg != null) {
                failMsg = errorMsg
                return false
            }

            // 清除 reader 缓存
            executor!!.executeCode("")

            val result = executor!!.executeCode("import uiautomator2 as u2")
            if (result.contains("ModuleNotFoundError:")) {
                println(result)
                failMsg = "please install uiautomator2 first, interpreter path: $path"
                println(failMsg)
                return false
            }

            val connectCode = if (serial == null) {
                "$objectName = u2.connect()"
            } else {
                "$objectName = u2.connect('$serial')"
            }
            val result2 = executor!!.executeCode(connectCode)

            if (result2.contains("NameError:") || result2.contains("状态异常")) {
                failMsg = result2
                println(failMsg)
                return false
            }
            true
        } catch (e: IOException) {
            e.printStackTrace()
            false
        }
    }

    fun getInitFailMsg(): String? = failMsg

    fun executeCode(code: String): String? {
        return executeCode(code, 0)
    }

    fun executeCode(code: String, indentLevel: Int): String? {
        if (executor == null) {
            return null
        }
        return try {
            executor!!.executeCode(code, indentLevel)
        } catch (e: IOException) {
            e.printStackTrace()
            null
        }
    }

    fun dumpUI(screenShotSaveFile: File, hierarchySaveFile: File) {
        // device.dump_ui(dir)
        executeCode("d.screenshot(r'" + screenShotSaveFile.absolutePath + "')")
        executeCode("with open(r'" + hierarchySaveFile.absolutePath + "', 'w', encoding=\"utf-8\") as f: f.write(d.dump_hierarchy())")
        executeCode("")
        executeCode("")
        executeCode("")
    }

    fun getNodeBounds(code: String): Rectangle? {
        val result = executeCode("print($code.info['bounds'] if $code.exists else None)")
        executeCode("")
        if (result == null || result.trim() == "None") {
            return null
        }
        println(result)
        // {'bottom': 1555, 'left': 296, 'right': 540, 'top': 1278}
        // 判断格式是否正确
        if (!result.contains("'left': ") || !result.contains("'top': ") ||
            !result.contains("'right': ") || !result.contains("'bottom': ")
        ) {
            return null
        }

        // 解析结果
        val left = NumberUtil.parseInt(result.split("'left': ")[1].split(",")[0])
        val top = NumberUtil.parseInt(result.split("'top': ")[1].split("}")[0])
        val right = NumberUtil.parseInt(result.split("'right': ")[1].split(",")[0])
        val bottom = NumberUtil.parseInt(result.split("'bottom': ")[1].split(",")[0])
        return Rectangle(left, top, right - left, bottom - top)
    }

    fun destroy() {
        try {
            executor?.destroy()
            executor = null
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}
